src/interfaces.ts

```diff
@@ -1,53 +1,86 @@
-import { IBackoffOpts } from 'backoff-web';
-
-export interface IServerOpts {
-  accessToken: string,  // genesyscloud auth token
-  environment: string,  // genesyscloud api environment
-  appVersion: string,   // version of app using the logging library
-  logTopic: string,     // all local logs will be prefixed by this
-  logLevel: LogLevels,  // logs at this level or high get sent to the server
-  uploadDebounceTime?: number,   // time to debounce logs uploads to the server
-}
-
-export interface IDeferred {
-  promise: Promise<any>;
-  reject: any;
-  resolve: any;
+export interface ILoggerConfig {
+  /** JWT access token to use in HTTP request */
+  accessToken: string;
+  /**
+   * url to send the logs to (note this needs to be the full URL)
+   * an HTTP `POST` request will be issued to this url
+   */
+  url: string;
+  /** version of app using the logging library */
+  appVersion: string;
+  /** all local logs will be prefixed by this. */
+  logTopic: string;
+  /** initialize server logging. defaults to `true` */
+  initializeServerLogging?: boolean;
+  /** logs at this level or high get sent to the server. defaults to 'info' */
+  logLevel?: ConfigLogLevel;
+  /** time to debounce logs uploads to the server */
+  uploadDebounceTime?: number;
 }
 
-export interface ISendLogState {
-  deferred: IDeferred;
-  request: ISendLogRequest;
-}
+export type LogLevel = 'log' | 'debug' | 'info' | 'warn' | 'error';
 
-export interface ISendLogRequest {
-  accessToken: string;
-  app: {
-    appId: string;
-    appVersion: string;
-  };
-  traces: ITrace[];
-}
+export type ConfigLogLevel = LogLevel | 'none';
 
 export interface ITrace {
   topic: string;
   level: string;
   message: string;
 }
 
-export interface RequestApiOptions {
-  accessToken?: string;
-  method?: string;
-  data?: any;
-  apiVersion?: string;
-  environment: string;
-  contentType?: string;
+export interface ILogMessage {
+  clientTime: string;
+  clientId: string;
+  message: string;
+  details?: IDetail | IDetail[];
 }
 
-export enum LogLevels {
-  log = 'log',
-  debug = 'debug',
-  info = 'info',
-  warn = 'warn',
-  error = 'error'
+export interface ILogBufferItem {
+  size: number;
+  traces: ITrace[];
 }
+
+export interface IDetailObject {
+  [k: string]: IDetailObject | string | number | boolean | undefined | null;
+  [k: number]: IDetailObject | string | number | boolean | undefined | null;
+};
+
+export type IDetail =
+  | Error
+  | string
+  | boolean
+  | number
+  | IDetailObject
+  | undefined
+  | null
+  | Event;
```

Comment:
This `IDetail` and `IDetailObject` is an idea from Josh. It forces you to think about what you are logging. However, it has draw backs. Read the next few lines to see what TS is happy with and what causes errors.